explorer/ui/src/components/Forms.tsx

```diff
@@ -35,46 +47,117 @@ interface FormProps {
   onSubmit?: () => void;
 }
 
+interface FormRowProps {
+  children?: any;
+  splits: number[];
+}
+
 interface FileUploaderProps {
   id: string;
   label: string;
+  width?: number;
   handleFileSelect: (e: React.ChangeEvent<HTMLInputElement>) => void;
 }
 
-function controlClass(props: FieldProps) {
+function controlClass(props: FieldProps<any>) {
+  if (props.readonly) {
+    return 'form-control';
+  }
   let validity =
-    props.valid == null || props.valid === undefined
-      ? ''
-      : props.valid
-      ? 'is-valid'
-      : 'is-invalid';
+    !props.fieldState.hasBeenValidated ?
+      ''
+      : props.fieldState.hasError
+      ? 'is-invalid'
+      : '';
   return ['form-control', validity].join(' ');
 }
 
 export const TextField = observer((props: TextProps) => {
+  let v: string;
+  let errorMsg = null;
+  if (props.fieldState instanceof FieldState) {
+    v = props.fieldState.value as string;
+    errorMsg = props.fieldState.error;
+  } else {
+    v = props.fieldState as string;
+  }
   let input = (
     <input
       className={controlClass(props)}
       id={props.id}
       type="text"
       placeholder={props.placeholder}
-      value={props.value || ''}
+      value={v}
       readOnly={props.readonly || false}
-      onChange={e => props.onChange!(e.target.value)}
+      onChange={e => {
+        if (!props.readonly || props.fieldState instanceof String) {
```

Comment:
```suggestion
        if (!(props.readonly || props.fieldState instanceof String)) {
```
Otherwise it can treat a `readonly` `String` as `FieldState`

Reply:
Thanks!